src/SignalR/common/Http.Connections.Common/src/NegotiateProtocol.cs

```diff
@@ -13,77 +15,166 @@ namespace Microsoft.AspNetCore.Http.Connections
         {
-            var textWriter = Utf8BufferTextWriter.Get(output);
-            try
+            var writer = new Utf8JsonWriter(output, new JsonWriterState(new JsonWriterOptions() { SkipValidation = true }));
+            writer.WriteStartObject();
+
+            if (!string.IsNullOrEmpty(response.Url))
+            {
+                writer.WriteString(UrlPropertyNameBytes, response.Url, escape: false);
+            }
+
+            if (!string.IsNullOrEmpty(response.AccessToken))
+            {
+                writer.WriteString(AccessTokenPropertyNameBytes, response.AccessToken, escape: false);
+            }
+
+            if (!string.IsNullOrEmpty(response.ConnectionId))
             {
-                using (var jsonWriter = JsonUtils.CreateJsonTextWriter(textWriter))
+                writer.WriteString(ConnectionIdPropertyNameBytes, response.ConnectionId, escape: false);
+            }
+
+            writer.WriteStartArray(AvailableTransportsPropertyNameBytes, escape: false);
+
+            if (response.AvailableTransports != null)
+            {
+                foreach (var availableTransport in response.AvailableTransports)
                 {
-                    jsonWriter.WriteStartObject();
+                    writer.WriteStartObject();
+                    writer.WriteString(TransportPropertyNameBytes, availableTransport.Transport, escape: false);
+                    writer.WriteStartArray(TransferFormatsPropertyNameBytes, escape: false);
 
-                    if (!string.IsNullOrEmpty(response.Url))
+                    if (availableTransport.TransferFormats != null)
                     {
-                        jsonWriter.WritePropertyName(UrlPropertyName);
-                        jsonWriter.WriteValue(response.Url);
+                        foreach (var transferFormat in availableTransport.TransferFormats)
+                        {
+                            writer.WriteStringValue(transferFormat, escape: false);
+                        }
                     }
 
-                    if (!string.IsNullOrEmpty(response.AccessToken))
-                    {
-                        jsonWriter.WritePropertyName(AccessTokenPropertyName);
-                        jsonWriter.WriteValue(response.AccessToken);
-                    }
+                    writer.WriteEndArray();
+                    writer.WriteEndObject();
+                }
+            }
 
-                    if (!string.IsNullOrEmpty(response.ConnectionId))
-                    {
-                        jsonWriter.WritePropertyName(ConnectionIdPropertyName);
-                        jsonWriter.WriteValue(response.ConnectionId);
-                    }
+            writer.WriteEndArray();
+            writer.WriteEndObject();
+
+            writer.Flush(isFinalBlock: true);
+        }
+
+        public static NegotiationResponse ParseResponse(byte[] content)
```

Comment:
I would be curious to see the performance of using the `JsonDocument` in this case compared to the `Utf8JsonReader`.

Reply:
I gave `JsonDocument` a quick try, although I'm definitely no expert on it's usage:

Method |       Mean |    Error |    StdDev |        Op/s |  Gen 0 | Allocated |
------------------------------------------------ |-----------:|---------:|----------:|------------:|-------:|----------:|
ParsingNegotiateResponseMessageSuccessForValid1 | 1,252.6 ns | 24.96 ns | 24.510 ns |   798,312.5 | 0.0057 |     528 B |
ParsingNegotiateResponseMessageSuccessForValid2 |   877.6 ns | 14.30 ns |  9.462 ns | 1,139,434.6 | 0.0038 |     368 B |
ParsingNegotiateResponseMessageSuccessForValid3 | 1,096.4 ns | 21.74 ns | 26.694 ns |   912,048.6 | 0.0057 |     536 B |
ParsingNegotiateResponseMessageSuccessForValid4 | 2,365.0 ns | 46.47 ns | 63.604 ns |   422,835.1 | 0.0076 |     944 B |
ParsingNegotiateResponseMessageSuccessForValid5 | 2,688.2 ns | 52.91 ns | 49.489 ns |   372,001.7 | 0.0114 |    1208 B |

Reply:
Thanks for giving it a try. Do you happen to have a commit with the code change to use `JsonDocument` that I can look at?

Reply:
I didn't save it, but it was fairly quick so I can make one for you.

Reply:
Take a look at https://github.com/aspnet/AspNetCore/compare/brecon/jsonNeg...brecon/jsonDoc?expand=1

This time I noticed the TryGetProperty method, so it was a little bit better.
The `ToArray` in `JsonDocument.Parse` should be avoidable if an overload was added to support `ReadOnlySpan<byte>`.

Side by side numbers:
**With JsonDocument**

Method |       Mean |    Error |   StdDev |        Op/s |  Gen 0 | Allocated |
------------------------------------------------ |-----------:|---------:|---------:|------------:|-------:|----------:|
 ParsingNegotiateResponseMessageSuccessForValid1 | 1,495.7 ns | 31.17 ns | 65.75 ns |   668,580.0 | 0.0019 |     264 B |
 ParsingNegotiateResponseMessageSuccessForValid2 | 1,070.6 ns | 22.78 ns | 43.34 ns |   934,042.2 | 0.0019 |     248 B |
 ParsingNegotiateResponseMessageSuccessForValid3 | 1,314.7 ns | 26.21 ns | 55.86 ns |   760,646.2 | 0.0019 |     304 B |
 ParsingNegotiateResponseMessageSuccessForValid4 | 2,482.5 ns | 49.28 ns | 88.86 ns |   402,823.8 | 0.0038 |     456 B |
 ParsingNegotiateResponseMessageSuccessForValid5 | 2,820.7 ns | 55.10 ns | 99.36 ns |   354,521.8 | 0.0038 |     720 B |

**With Utf8JsonReader**

Method |       Mean |     Error |   StdDev |        Op/s |  Gen 0 | Allocated |
------------------------------------------------ |-----------:|----------:|---------:|------------:|-------:|----------:|
 ParsingNegotiateResponseMessageSuccessForValid1 |   404.7 ns |  8.122 ns | 13.79 ns | 2,470,926.1 | 0.0010 |     120 B |
 ParsingNegotiateResponseMessageSuccessForValid2 |   305.5 ns |  6.168 ns | 15.13 ns | 3,273,484.3 | 0.0010 |     120 B |
 ParsingNegotiateResponseMessageSuccessForValid3 |   461.3 ns |  9.164 ns | 18.09 ns | 2,167,912.7 | 0.0014 |     152 B |
 ParsingNegotiateResponseMessageSuccessForValid4 |   752.4 ns | 14.997 ns | 29.95 ns | 1,329,015.3 | 0.0029 |     272 B |
 ParsingNegotiateResponseMessageSuccessForValid5 | 1,101.3 ns | 22.009 ns | 60.25 ns |   908,003.1 | 0.0038 |     480 B |

Reply:
The `ToArray()` isn't great :(

Can you try copying it to an array pool-rented array instead and share how much that helps?
```C#
// BEFORE:
using (var doc = JsonDocument.Parse(content.ToArray(), readerOptions: default))
```
```C#
// AFTER:
byte[] pooledBuffer = ArrayPool<byte>.Shared.Rent(content.Length);
content.CopyTo(pooledBuffer);

using (var doc = JsonDocument.Parse(pooledBuffer, readerOptions: default))
{
   ...
}
ArrayPool<byte>.Shared.Return(pooledBuffer);

```

I expected a ~2x regression between reader and document, not 3-4x.

Reply:
Better yet, can we change the API signature to accept `ReadOnlyMemory<byte>` instead to avoid having to copy data at all? It looks like the only caller passes in a byte[] anyway:
```C#
public static NegotiationResponse ParseResponse(ReadOnlySpan<byte> content)
```

https://github.com/aspnet/AspNetCore/pull/6977/files#diff-041ab9d757596cece2ef39f92ede7e4dR443

Reply:
Thanks for sharing the numbers. They provide useful insight.